Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs

```diff
@@ -0,0 +1,211 @@
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.CatalogModule.Data.ExportImport
+{
+    public class CatalogFullExportPagedDataSource : IPagedDataSource
+    {
+        private class FetchResult
+        {
+            public int TotalCount;
+            public IEnumerable<IExportable> Result = Array.Empty<IExportable>();
+        }
+
+        private class ExportDataSourceState
+        {
+            public FetchResult FetchResult;
+            public SearchCriteriaBase SearchCriteria;
+            public Func<ExportDataSourceState, Task<FetchResult>> FetchFunc;
+
+            public ExportDataSourceState(SearchCriteriaBase searchCriteria, Func<ExportDataSourceState, Task<FetchResult>> fetchFunc)
+            {
+                SearchCriteria = searchCriteria;
+                FetchFunc = fetchFunc;
+            }
+
+            public void SetFetchResult<T>(GenericSearchResult<T> searchResult) where T : Entity
+            {
+                FetchResult = new FetchResult()
+                {
+                    TotalCount = searchResult.TotalCount,
+                    Result = searchResult.Results.Cast<IExportable>()
+                };
+            }
+        }
+
+        private readonly ICatalogSearchService _catalogSearchService;
+        private readonly IProductSearchService _productSearchService;
+        private readonly ICategorySearchService _categorySearchService;
+        private readonly IPropertySearchService _propertySearchService;
+        private readonly IProperyDictionaryItemSearchService _propertyDictionarySearchService;
+
+        public CatalogFullExportPagedDataSource(ICatalogService catalogService, ICatalogSearchService catalogSearchService, IProductSearchService productSearchService, ICategorySearchService categorySearchService, ICategoryService categoryService,
+                                  IItemService itemService, IPropertyService propertyService, IPropertySearchService propertySearchService, IProperyDictionaryItemSearchService propertyDictionarySearchService,
+                                  IProperyDictionaryItemService propertyDictionaryService, IBlobStorageProvider blobStorageProvider, IAssociationService associationService)
+        {
+            _productSearchService = productSearchService;
+            _categorySearchService = categorySearchService;
+            _propertySearchService = propertySearchService;
+            _propertyDictionarySearchService = propertyDictionarySearchService;
+            _catalogSearchService = catalogSearchService;
+        }
+
+        public int TotalCount { get; set; }
+        public int CurrentPageNumber { get; protected set; }
+        public int PageSize { get; set; } = 50;
+
+        public IEnumerable<IExportable> Items { get; protected set; }
+
+        private bool _statesInitialized = false;
+
+        private readonly ExportDataSourceState[] _exportDataSourceStates = new ExportDataSourceState[5];
```

Comment:
Could you use array for this? Initialization will become easier.

Reply:
Removed

Reply:
Here is an array already. Let's discuss on monday.